### PR TITLE
Fix issue 57762

### LINKED ISSUE #57762: rustc 1.32 using system LLVM crashes while building the resvg crate
Initial issue: https://github.com/RazrFalcon/resvg/issues/110

```
   Compiling resvg-capi v0.5.0 (/home/user/.cache/pikaur/build/resvg-git/src/resvg/capi)
error: Could not compile `resvg-capi`.

Caused by:
  process didn't exit successfully: `rustc --crate-name resvg capi/src/lib.rs --color always --crate-type cdylib --crate-type staticlib --emit=dep-info,link -C opt-level=3 -C lto --cfg 'feature="cairo-backend"' --cfg 'feature="cairo-sys-rs"' --cfg 'feature="glib"' --cfg 'feature="qt-backend"' --cfg 'feature="resvg"' -C metadata=34044ee5526873f1 --out-dir /home/user/.cache/pikaur/build/resvg-git/src/resvg/target/release/deps -L dependency=/home/user/.cache/pikaur/build/resvg-git/src/resvg/target/release/deps --extern cairo_sys=/home/user/.cache/pikaur/build/resvg-git/src/resvg/target/release/deps/libcairo_sys-a11059b04a916894.rlib --extern fern=/home/user/.cache/pikaur/build/resvg-git/src/resvg/target/release/deps/libfern-f758ae3d5f3ac504.rlib --extern glib=/home/user/.cache/pikaur/build/resvg-git/src/resvg/target/release/deps/libglib-42c67ca816dc468b.rlib --extern log=/home/user/.cache/pikaur/build/resvg-git/src/resvg/target/release/deps/liblog-4fed47ee5064f102.rlib --extern resvg=/home/user/.cache/pikaur/build/resvg-git/src/resvg/target/release/deps/libresvg-9bfdf02a73881248.rlib -L native=/home/user/.cache/pikaur/build/resvg-git/src/resvg/target/release/build/resvg-qt-529ad70e8602e985/out` (signal: 11, SIGSEGV: invalid memory reference)
```

Steps to reproduce:
```
sudo pacman -S rust
git clone https://github.com/RazrFalcon/resvg.git
cd resvg/capi
cargo build --release
```

The master branch can be built without rustc crashing on Gentoo and on Travis (Ubuntu 14.04.5 LTS).

**UPD:** I've tried to build it too and I have the same error. Looks like something wrong with the Arch's rustc. Because if we use the one installed via `rustup` - everything is ok.

**UPD:** it can be fixed by disabling the **LTO**, which is enabled by default for `resvg`.

### PR DESCRIPTION
**Pull request by tromey**
against a stock LLVM 7.  LLVM 7 was released without a necessary fix
for a bug in the DWARF discriminant code.

This patch changes rustc to use the fallback mode on (non-Rust) LLVM 7.

Closes #57762

### CODE REVIEW

**Comment by tromey** [comment]
Someone the commit message was munged.

**Comment by tromey** [comment]
*somehow.  apparently I cannot type today.  anyway, pushed a better message.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 1]
```diff
@@ -1164,7 +1164,11 @@ fn use_enum_fallback(cx: &CodegenCx) -> bool {
     // On MSVC we have to use the fallback mode, because LLVM doesn't
     // lower variant parts to PDB.
     return cx.sess().target.target.options.is_like_msvc
-        || llvm_util::get_major_version() < 7;
+        || llvm_util::get_major_version() < 7
+        // LLVM version 7 did not release with an important bug fix;
+        // but the required patch is in the equivalent Rust LLVM.
+        // See https://github.com/rust-lang/rust/issues/57762.
+        || (llvm_util::get_major_version() == 7 && unsafe { !llvm::LLVMRustIsRustLLVM() });
```
Comment (nikic) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: It's enough to only check `llvm_util::get_major_version() < 8` here. Rust's LLVM fork will report version 8 (or, in the future, higher).

**Pull request merged**